### PR TITLE
7538 - Validate cookie when restoring the app from the bfcache 

### PR DESCRIPTION
**Pull request by m5r**
# Description

The bug had to do with the browser's bfcache, see [related documentation](https://web.dev/bfcache/)  
Essentially, it restored the page and its state from memory without even hitting the network. Showing a half logged in/half logged out state when coming back to the app from the login page with the browser's back button.

With this PR, when the app is restored from the bfcache, the session is validated once more to make sure the user doesn't get stuck in an impossible state.

medic/cht-core#7538

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Review by dianabarsan — COMMENTED** [review]
Nice work! 
Three things: 
1. Please add an e2e test for this
2. Please add the same logic in the admin app
3. This happened a few times, where I loaded the login page when doing a back. 
The spinner appears immediately (without me clicking anything) and stays there forever, and the button is disabled, just like cache was also used to load the page. 
image

File: tests/wdio.conf.js [hunk 1]
```diff
@@ -78,7 +79,7 @@ const baseConfig = {
     browserName: 'chrome',
     acceptInsecureCerts: true,
     'goog:chromeOptions': {
-      args: ['--headless', '--disable-gpu', '--deny-permission-prompts']
+      args: [/*'--headless',*/ '--disable-gpu', '--deny-permission-prompts']
```
Comment (latin-panda) [inline tests/wdio.conf.js:82]: just curious, why are we commenting the `--headless`?

File: tests/wdio.conf.js [hunk 2]
```diff
@@ -40,7 +40,8 @@ const baseConfig = {
   // will be called from there.
   //
   specs: [
-    './tests/e2e/**/*.wdio-spec.js'
+    // './tests/e2e/**/*.wdio-spec.js',
```
Comment (latin-panda) [inline tests/wdio.conf.js:43]: Do we need this commented line?

File: tests/wdio.conf.js (hunk 1 above)
Comment (m5r) [reply tests/wdio.conf.js:82]: both changes in `wdio.conf.js` are temporary, I'm trying to reproduce a bug in the CI that I have locally, sorry for the disruption

File: admin/src/js/services/session.js [hunk 3]
```diff
@@ -114,6 +114,8 @@ const _ = require('lodash/core');
 
         init: checkCurrentSession,
 
+        checkCurrentSession: checkCurrentSession,
```
Comment (m5r) [inline admin/src/js/services/session.js:117]: I've defined another export here named `checkCurrentSession` so it conveys better the intent when calling `Session.checkCurrentSession()`
I couldn't find any `Session.init()` call in the admin app so I wasn't sure if I could simply rename that export

File: admin/src/js/services/db.js [hunk 4]
```diff
@@ -22,6 +22,7 @@ angular.module('inboxServices').factory('DB',
     const isOnlineOnly = Session.isOnlineOnly();
 
     const getUsername = remote => {
+      Session.checkCurrentSession();
```
Comment (m5r) [inline admin/src/js/services/db.js:25]: This gets called immediately when coming back to the admin app with the back button. It strangely gets called before the `pageshow` event handler and this function throwing an error results in the event handler not being called 🤨

But hey, no need for an extra event listener at least :)

**Comment by m5r** [comment]
@dianabarsan @latin-panda could you review this PR once more? I've fixed the problem in the admin app, on the login page, and I've added e2e tests for those 3 scenarios

File: tests/wdio.conf.js (hunk 2 above)
Comment (latin-panda) [inline tests/wdio.conf.js:43]: Remember to revert this one when you've finish testing :)

Comment (m5r) [reply tests/wdio.conf.js]: Oops, thanks!

**Review by dianabarsan — COMMENTED** [review]
Nice! Just one question inline.

File: admin/src/js/services/session.js (hunk 3 above)
Comment (latin-panda) [reply admin/src/js/services/session.js:117]: Yes, same, I didn't find any usage of `Session.init()`, I'd remove it and just have `Session.checkCurrentSession()`

File: webapp/src/ts/services/session.service.ts [hunk 5]
```diff
@@ -73,6 +78,13 @@ export class SessionService {
     if (!userCtx || !userCtx.name) {
       return this.logout();
     }
+
+    window.addEventListener('pageshow', (event) => {
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply webapp/src/ts/services/session.service.ts:82]: I edited the above comment (better code). I was sleepy when I first submitted, sorry.

[the remark this answers is not in the document]
Comment (m5r) [reply webapp/src/ts/services/session.service.ts]: Oh, thanks, that's much cleaner indeed! I bet Angular also has some logic to remove event listeners when appropriate.  
And take care of your sleep :zzz:

**Review by dianabarsan — APPROVED** [review]
Nice work!

**Pull request merged**